### PR TITLE
Toreview

### PR DESCRIPTION
**Pull request by TheSpaceSheep**
## Description
To manage memory when dealing with very large input spaces, this request adds the functionality of sampling a dataset of chosen length from the input space, instead of building the entire dataset.

### CODE REVIEW

File: egg/zoo/compo_vs_generalization/data.py [hunk 1]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
     return train, test
 
 
+def build_test_set(n_attributes, n_values, size):
+    """
+    Samples {size} elements from the dataset, with no possible repeat
+    """
+    max_size = 10e6  # TODO: find better value for this empirically
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: This is a good idea but it probably depends on the n_attr/n_vals combinations. One idea is to check for the probability of sampling a duplicate as the `size`th element as function of size, n_attr and n_vals and raise a warning when this probability is greater than a certain threshold (like 0.1).

This link might contain some indication towards the exact derivation: https://math.stackexchange.com/questions/1056540/sampling-with-replacement-expected-number-of-duplicates-triplicates-n-t

File: egg/zoo/compo_vs_generalization/data.py [hunk 2]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+    if size > max_size:
+        print(f"Warning : trying to build a test set of size greater than {max_size} which might take a long time")
+
+    # keep data in a tree for faster lookup
+    data_tree = BinarySearchTree()
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: It is probably better to use the native `set()` which might be more efficient.

File: egg/zoo/compo_vs_generalization/data.py [hunk 3]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+
+    data = []
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: You probably don't need two data structures. You can simply convert `data_tree` to a list afterward

File: egg/zoo/compo_vs_generalization/data.py [hunk 4]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+    i = 0
+    while i < size:
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: You can directly test for `len(data) < size`, this way no need to rely on an auxiliary variable i

File: egg/zoo/compo_vs_generalization/data.py [hunk 5]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+        sample = tuple([random.randint(0, n_values-1) for _ in range(n_attributes)])
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: For random sampling, I suggest sticking to either np.random or torch.random. The reason is that

1. Faster parallel sampling (you don't need to run a for loop, you can do np.randint(n_values, size=(n_attributes,)) or similar in pytorch)
2. It will be easier to handle random seeds and centralize the choice of RNG

That particular point is very important. In particular, the way the function is written right now, it will return a different dataset every time it's called. Since this is generating our dataset, we want to be able to make it return a consistent result.

A solution for this is to have a random number generator as an argument of this function. This allows you to define an RNG with a fixed seed outside of the function, and use it to generate the data.

Here is an example  in pytorch:

```python
# Define rng
data_rng = th.Generator()
# Set seed manually
data_rng.manual_seed(394639)
# Define function with seed
def make_dataset(..., rng):
    # [...]
    x = rng.randint(...)
    # [...]
# Call function with RNG
data = make_dataset(..., rng=data_rng)
```

More info about seeds and reproducibility: 
- seeding: https://pytorch.org/docs/stable/notes/randomness.html
- RNGs: https://pytorch.org/docs/stable/generated/torch.Generator.html (basically this can be used to call any random sampling function in pytorch with a dedicated RNG)

File: egg/zoo/compo_vs_generalization/data.py [hunk 6]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+
+def build_datasets(n_attributes, n_values, train_size, test_size, validation_size):
+    """
+    Returns the train, test and validation sets by sampling the data.
+    Samples can be repeated in the training set but not in the testing set
+    Also there is no overlap between train and test set.
+    """
+    # prevent infinite loops
+    assert test_size < train_size and validation_size < train_size, 
+      "please set the --train_size option to be (much) larger than test and validation sizes."
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: This shouldn't really be a requirement

File: egg/zoo/compo_vs_generalization/data.py [hunk 7]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+
+    # create test and validation sets simultaneously to prevent repeats
+    all_test_data = build_test_set(n_attributes, n_values, size=test_size+validation_size)
+    test_set = all_test_data[:test_size]
+    validation_set = all_test_data[test_size:]
+
+    # build tree to speed up the search
+    test_set_tree = BinarySearchTree()
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: Similar comment as test_set: you can use `set()`

File: egg/zoo/compo_vs_generalization/data.py [hunk 8]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
     return train, test
 
 
+def build_test_set(n_attributes, n_values, size):
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: You could probably reuse this function (w/ a more generic name) to create the train set also. Just add an optional argument (eg. `data_to_exclude`) which is a python `set` containing samples to exclude from sampling (eg. test and/or validation data)

File: egg/zoo/compo_vs_generalization/data.py [hunk 9]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+    for sample in test_set:
+        test_set_tree.insert(sample)
+
+    train_set = []
+    i = 0
+    while i < train_size:
+        sample = tuple([random.randint(0, n_values-1) for _ in range(n_attributes)])
+        if sample not in test_set_tree:
+           train_set.append(sample)
+           i += 1
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: See my earlier comment: this looks very similar to the function you wrote for the test set: you could probably fold that into a more general function for generating datasets. Also, you might want to keep the option to have training sets without duplicates (which I think should be on for now).

File: egg/zoo/compo_vs_generalization/data.py [hunk 10]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+
+    for s in train_set:
+        assert s not in test_set, "Found overlap between train and test set"
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: Good to have this check but it might slow things down for bigger datasets. If this becomes an issue later on you could take this line out (and maybe put it in a unit test if you want)

File: egg/zoo/compo_vs_generalization/train.py [hunk 11]
```diff
@@ -90,6 +91,30 @@ def get_params(params):
         action="store_true",
         help="Whether the messages are variable or fixed length",
     )
+    parser.add_argument(
+        "--full_data",
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py]: This name is not very self explanatory

File: egg/zoo/compo_vs_generalization/train.py [hunk 12]
```diff
@@ -90,6 +91,30 @@ def get_params(params):
...
+    parser.add_argument(
+        "--train_size",
+        type=int,
+        default="100_000",
+        help="Size of the training set",
+    )
+    parser.add_argument(
+        "--test_size",
+        type=int,
+        default="500",
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py]: 1000 is probably a better default

File: egg/zoo/compo_vs_generalization/train.py [hunk 13]
```diff
@@ -90,6 +91,30 @@ def get_params(params):
...
+        help="Size of the testing set",
+    )
+    parser.add_argument(
+        "--validation_size",
+        type=int,
+        default="500",
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py]: Same

File: egg/zoo/compo_vs_generalization/train.py [hunk 14]
```diff
@@ -290,13 +334,13 @@ def main(params):
...
-            DiffLoss(opts.n_attributes, opts.n_values, generalization=True),
-        )
-    )
+    #    loaders.append(
+    #        (
+    #            "generalization hold out",
+    #            generalization_holdout_loader,
+    #            DiffLoss(opts.n_attributes, opts.n_values, generalization=True),
+    #        )
+    #    )
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py]: Why is this commented out?

File: egg/zoo/compo_vs_generalization/data.py (hunk 8 above)
Comment (pmichel31415) [reply egg/zoo/compo_vs_generalization/data.py]: Also you might add type annotation as well

File: egg/zoo/compo_vs_generalization/bst.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        if value > self.value:
+            if self.right is None:
+                self.right = BinarySearchTree(value)
+            else:
+                self.right.insert(value)
+        else:
+            if self.left is None:
+                self.left = BinarySearchTree(value)
+            else:
+                self.left.insert(value)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/bst.py]: So this is just an anecdotal comment since I think you should replace this data structure with the native python hashmap (`set()`) but I think a drawback with this implementation is that the tree might not be balanced: if you insert all the elements in decreasing (or increasing) order, the tree will be right (or left) branching and so lookup will be O(n).

Also, keep in mind you are using this data structure with tuples, so you should be careful how python handles comparison with these (for the record I think it is by lexicographic order).

All the more reasons why it's simpler to stick with the native implementation :)

File: egg/zoo/compo_vs_generalization/data.py [hunk 16]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+    data_tree = BinarySearchTree()
+
+    data = []
+    i = 0
+    while i < size:
+        sample = tuple([random.randint(0, n_values-1) for _ in range(n_attributes)])
+        if sample not in data_tree:
+           data_tree.insert(sample)
+           data.append(sample)
+           i += 1
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: Might be worth adding a couple comments here to explain what's going on

File: egg/zoo/compo_vs_generalization/data.py [hunk 17]
```diff
@@ -120,6 +124,66 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+
+    assert len(data) == len(set(data)), "Found duplicates in test set"
+
+    return data
+
+
+def build_datasets(n_attributes, n_values, train_size, test_size, validation_size):
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: Probably need type annotations (also return type)

**Review by pmichel31415 — CHANGES_REQUESTED** [review]
Good work! I left a few comments. My main points are regarding:

- Random sampling
- The choice of data structure to check for duplicates
- Structuring of the code to avoid repetition

File: egg/zoo/compo_vs_generalization/data.py [hunk 18]
```diff
@@ -120,6 +124,65 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
     return train, test
 
 
+def build_random_dataset(n_attributes: int,
+                         n_values: int,
+                         size: int,
+                         data_to_exclude: set=set(),
+                         allow_duplicates: bool=False,
+                         rng=None
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: You can even annotate this as `th.Generator`

File: egg/zoo/compo_vs_generalization/data.py [hunk 19]
```diff
@@ -120,6 +124,65 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
     return train, test
 
 
+def build_random_dataset(n_attributes: int,
+                         n_values: int,
+                         size: int,
+                         data_to_exclude: set=set(),
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: Be careful with mutable defaults. You should use an immutable default (None) and then handle the creation of a new set.

Also, for arguments with default values you can use the `Optional` typing:

```python
def function(x: Optional[int] = 0):
    pass
```

File: egg/zoo/compo_vs_generalization/data.py [hunk 20]
```diff
@@ -120,6 +124,65 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+                         allow_duplicates: bool=False,
+                         rng=None
+                         ) -> list:
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: You can even specify the type of the list with `typing.List`. For instance a list of ints is annotated `List[int]`. Here it would be `List[Tuple]`

File: egg/zoo/compo_vs_generalization/data.py [hunk 21]
```diff
@@ -120,6 +124,65 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+    with no overlap with the set {data_to_exclude},
+    """
+    if n_values**n_attributes / size < 10e4 and allow_duplicates:
+        print(f"Warning : Building a dataset that probably contains duplicates")
+
+    # use a list if duplicates are allowed (uses less memory than set)
+    data = list() if allow_duplicates else set()
+    
+    while len(data) < size:
+        sample = tuple(torch.randint(n_values-1, size=(n_attributes,), generator=rng).tolist())
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: Note that the high value is exclusive in torch.randint (https://pytorch.org/docs/stable/generated/torch.randint.html , same in other libraries). So this will only return numbers within 0 and n_values-2

File: egg/zoo/compo_vs_generalization/data.py [hunk 22]
```diff
@@ -120,6 +124,65 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
...
+            size=test_size+validation_size,
+            allow_duplicates=False,
+            rng=rng)
+    test_set = all_test_data[:test_size]
+    validation_set = all_test_data[test_size:]
+
+    train_set = build_random_dataset(n_attributes,
+            n_values,
+            size=train_size,
+            data_to_exclude=set(test_set),
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: This doesn't check for overlap with the validation set (which we also want to avoid, otherwise we will over-estimate our validation accuracy). You can join the two sets with `set(test_set) + set(validation_set)` I think (or use `all_test_data`)

File: egg/zoo/compo_vs_generalization/train.py [hunk 23]
```diff
@@ -90,6 +91,36 @@ def get_params(params):
...
+    parser.add_argument(
+        "--validation_size",
+        type=int,
+        default="1000",
+        help="Size of the validation set",
+    )
+    parser.add_argument(
+        "--data_seed",
+        type=int,
+        default=0,
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py]: I think that 0 means that the seed will be drawn at random (not 100% sure but it is the case in some libraries). I suggest using a nonzero seed just to be safe

File: egg/zoo/compo_vs_generalization/train.py [hunk 24]
```diff
@@ -201,32 +232,60 @@ def main(params):
     device = opts.device
     print(opts)
 
-    full_data = enumerate_attribute_value(opts.n_attributes, opts.n_values)
-    if opts.density_data > 0:
-        sampled_data = select_subset_V2(
-            full_data, opts.density_data, opts.n_attributes, opts.n_values
+    if opts.build_full_dataset:
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py:235]: Put a warning that this is deprecated

File: egg/zoo/compo_vs_generalization/train.py [hunk 25]
```diff
@@ -201,32 +232,60 @@ def main(params):
...
+        
+        train, validation, uniform_holdout = [
+            one_hotify(x, opts.n_attributes, opts.n_values)
+            for x in [train, validation, uniform_holdout]
+        ]
 
-    train, validation = ScaledDataset(train, opts.data_scaler), ScaledDataset(train, 1)
+        train = ScaledDataset(train, opts.data_scaler)
+        validation = ScaledDataset(validation, 1)
+        uniform_holdout = ScaledDataset(uniform_holdout)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py]: rename test

**Review by pmichel31415 — CHANGES_REQUESTED** [review]

File: egg/zoo/compo_vs_generalization/train.py [hunk 26]
```diff
@@ -201,32 +232,57 @@ def main(params):
...
-    ]
+        train, validation, test = [
+            one_hotify(x, opts.n_attributes, opts.n_values)
+            for x in [train, validation, test]
+        ]
 
-    train, validation = ScaledDataset(train, opts.data_scaler), ScaledDataset(train, 1)
+        train = ScaledDataset(train, opts.data_scaler)
+        validation = ScaledDataset(validation, 1)
+        test = ScaledDataset(test)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py:282]: You probably don't need this wrapper here since we're not scaling the test and validation data

**Review by pmichel31415 — COMMENTED** [review]
LGTM!
